### ISSUE TITLE
Adopting a Progressive Photorealistic Global Illumination in Three.JS

### ISSUE THREAD

**Issue by bhouston** [mrdoob/three.js#14051]
There has recently been a bunch of work towards progressive photo-realistic global illumination in WebGL.  The best example I've seen of this is https://cl3ver.com's renderer.  There is some overlap between this and the high end rendering that Octane and other photorealistic GPU renderers do. I think it is time to add this to Three.JS.

This is a meta-task that describes what needs to be done (this is in the spirit of the PBR meta task from 2015: https://github.com/mrdoob/three.js/issues/5847)

The main components of this approach would be:

Temporal Reprojection Anti-Aliasing
https://github.com/mrdoob/three.js/issues/14050

Progressive Global Illumination via Instant Radiosity/Virtual Point Lights
https://github.com/mrdoob/three.js/issues/14047

Progressive Soft Shadows via Light Source Sampling
https://github.com/mrdoob/three.js/issues/14048

Screen-space Reflections
https://github.com/mrdoob/three.js/issues/8248

**Comment by gkjohnson** [comment]
Activision's [Ground Truth Ambient Occlusion](https://research.activision.com/t5/Publications/Practical-Real-Time-Strategies-for-Accurate-Indirect-Occlusion/ba-p/10002180) and [Unigine's SSRTGI](https://80.lv/articles/ssrtgi-toughest-challenge-in-real-time-3d/) both look really impressive and would be great to have available.

Is the assumption that these would come after something like multiple render textures or deferred rendering is available in THREE?

**Comment by bhouston** [comment]
Real-time light probes can be either diffuse or specular.  Diffuse light probes are made obsolete by instant radiosity.  Specular light probes are made obsolete via SSR.

That said, this solution is for **progressive** and very high quality result.  If you want an instantaneous and a bit lower quality solution you should use light probes.

**Comment by EliasHasle** [comment]
How about this? https://github.com/erichlof/THREE.js-PathTracing-Renderer

**Comment by elfrank** [comment]
Hi,

At [HOVER](https://hover.to), we recently open-sourced our Ray Tracing Renderer. It's a **drop-in** replacement for the WebGLRenderer in three.js.

![preview](https://user-images.githubusercontent.com/230761/64495340-8f4cca80-d24d-11e9-938d-2d0ce3dd7a0a.jpg)

**Links**
* [Repository](https://github.com/hoverinc/ray-tracing-renderer/)
* [Demo: WebGLRenderer vs RayTracingRenderer](https://hover-demos.s3.amazonaws.com/ray-tracing-renderer/readme-example/index.html)

**Note**
* You need WebGL 2.

**Comment by EliasHasle** [comment]
@elfrank Nice demo! It takes a while to get a nicer render with the raytracer than with the WebGLRenderer, though. Also, the WebGLRenderer is initialized without antialiasing, which would significantly improve the result.

**Comment by mrdoob** [comment]
@elfrank That looks great!

Things render pretty glitchy on ChromeOS/Pixelbook though 😶

![Screenshot 2019-09-09 at 23 32 47](https://user-images.githubusercontent.com/97088/64589533-2dc65200-d35a-11e9-8957-ba949d9e1587.png)

**Comment by looeee** [comment]
> That said, this solution is for progressive and very high quality result.

Does progressive in this context mean something like what sketchfab does? It takes maybe 0.5 seconds or less there to see the high quality result after moving the camera. That's something I'd love to see made possible in three.js. 

Or is more like the raytracer (or maybe a bit faster) where it takes several seconds to see a good result even on fast hardware?

**Comment by mrdoob** [comment]
>Does progressive in this context mean something like what sketchfab does? It takes maybe 0.5 seconds or less there to see the high quality result after moving the camera. 

That's what this example does (minus the improved shadows):
https://threejs.org/examples/#webgl_postprocessing_taa

**Comment by looeee** [comment]
I think sketchfab does more than just disable AA while the camera is moving though. It seems like the reduce the resolution and possibly reduce texture quality too, and perhaps some other things that are less obvious (maybe the use progressive shadows and lighting like @bhouston is suggesting here). 

I tried a quick tests to reduce resolution while `OrbitControls` is engaged but I couldn't get decent results or any noticeable performance gain, even in a fill rate limited scene.

**Comment by gkjohnson** [comment]
> I think sketchfab does more than just disable AA while the camera is moving though.

Some of this has already been mentioned but here's what I noticed with sketchfabs renderer if it's worth noting:

It's most pronounced in [this bacteria model](https://sketchfab.com/3d-models/tetanus-bacterium-clostridium-tetani-438a7e0a40964b48b87fdfa91ad2a5b5) but you can see that it renders transparency using some kind of noise distribution that then resolves to a smooth look. Presumably this is to help alleviate artifacts associated with transparent object render order.

It also looks like some kind of sharpen filter gets applied to the image before the accumulation kicks in but I'm not sure why. It's pretty easy to see in [this car model](https://sketchfab.com/3d-models/car-bus-bagel-concept-f1d505c90f5a474783cc6c2c74fd7a6b) when you zoom in and out (notice the light outlines around the dark edges of the model). Maybe someone else has an idea as to what this is for?

As was mentioned shadow map camera is subpixel jittered (much like the TAA approach, I think) to smooth shadow resolution artifacts. This can be seen pretty easily in [this cyborg boy model](https://sketchfab.com/3d-models/cyborg-boy-25e0e77ab56140be9fa6cb65cb4de96f) if you pause the animation and zoom into the shirt on the right arm.

Of course there's some really nice high quality post processing going on, too.

I'm curious to what the plan should be for adding these types features into the library. As far as I'm concerned it's already possible to achieve all of this but maybe it's worth adding some facilities to make it easier to achieve? And an example that glues all the necessary pieces together?

**Comment by bhouston** [comment]
I've analyzed Sketchfab.  It has a series of options that you can turn on and off so exactly what it does on a specific scene is dependent upon that.  But it has these capabilities:

- Sketchfab supports a series of optional and parameter post processing filters like sharpen, color correct, vignetting, sepia, chromatic aberration, etc.

- TRAA - temporal reproduction anti-aliasing.  It uses a velocity buffer of the scene to allow one to to TAA (https://threejs.org/examples/#webgl_postprocessing_taa) like effects but with a moving camera or deforming object.  I have code for this and I want our team here (ThreeKit) to contribute it back to Three.js, just it is lower priority than contributing back enhanced PBR materials (e.g. https://github.com/mrdoob/three.js/issues/16977 )

- Shadow camera jitter when the scene is static.  This smoothes out the shadows.  One should jitter the shadow cameras within the light volume for correct soft shadows.  When moving, one should position the shadow camera at the center of the light volume or at least the functional equivalent.

- Stochastic transparency.  This is what we see in the bacteria model I believe: http://luebke.us/publications/StochTransp-slides.pdf  I think Stochastic transparency can resolve to a good solution, but it looks like crap in dynamic scenes.  I am unsure of its value.

(An aside: Alternatively, we (Threekit) have implemented order independent transparency based on the McGuire model, http://casual-effects.blogspot.com/2014/03/weighted-blended-order-independent.html  But it is also highly problematic - it is dependent upon a lot of parameter tweaking and it is slow and needs floating point buffers.  I would recommend against implement this algorithm in Three.JS as well.

I am unsure of the best approach to OIT - both Stochastic and the McGuire OIT method have significant downsides.)

A suggested plan forward:

- I think we should add shadow camera jitter within the shadow emission volumes.  It should be something one can turn on and then on each frame it jitters within that region.  Then users can enable it during a TAA refinement phase.  It works exceptionally well with the TAA post effect I wrote.

- We should add TRAA and we can contribute that code.  Maybe we can contribute broken code ripped out of our custom Three.JS implementation and a hero can adapt it to the mainstream Three.JS?  (Interesting fact, you can not easily combine jitter shadow cameras with TRAA because TRAA uses a luminance discriminator to determine if things will blend across frames, and well shadow boundaries have significant luminance contrast -- but maybe someone can figure out a solution to this.)

Some people have combined TRAA with stochastic transparency. https://casual-effects.com/research/Wyman2017Improved/Wyman2017Improved.pdf https://developer.download.nvidia.com/assets/gameworks/downloads/regular/GDC17/RealTimeRenderingAdvances_HashedAlphaTesting_GDC2017_FINAL.pdf  Maybe the game INSIDE used this approach -- one slide says stochastic everywhere: http://twvideo01.ubm-us.net/o1/vault/gdc2016/Presentations/Pedersen_LasseJonFuglsang_TemporalReprojectionAntiAliasing.pdf

**Comment by elfrank** [comment]
> @elfrank Nice demo! It takes a while to get a nicer render with the raytracer than with the WebGLRenderer, though.

Yes, we use path tracing, which could be broadly bucketed as a "progressive photorealistic Global Illumination" technique. We are looking into faster convergence. Path tracing is expensive but there is a good amount of recent research on the matter.


> Also, the WebGLRenderer is initialized without antialiasing, which would significantly improve the result.

@EliasHasle Noted! Thanks.


> Things render pretty glitchy on ChromeOS/Pixelbook though 😶

That's unfortunate @mrdoob . We have experienced a few driver bugs/glitches in some devices. It's nearly possible to account for all scenarios without having access to the hardware. With that said, Pull requests are welcomed! Thoughts @jaxry ?


> Or is more like the raytracer (or maybe a bit faster) where it takes several seconds to see a good result even on fast hardware?

@looeee I highly recommend this video on Path Tracing: https://www.youtube.com/watch?v=frLwRLS_ZR0&t=11s
As you pointed out, the technique is inherently expensive. Hardware keeps getting faster and that helps a lot.

**Comment by zalo** [comment]
(Jumping to this thread for minor updates until (maybe) I get this clean and pretty enough for a PR.)

I've been implementing [Approximate Radiosity using Stochastic Depth Buffering](http://xevius.org/papers/approxradpaper2010.pdf), which is a version of the recent Shadow PR that tunnels through surfaces to cast rays throughout the scene.  This should enable reasonably fast, physically correct global illumination and area lighting (without path tracing or acceleration structures).

I don't have the indirect accumulation function correct yet, but here is an example scene that sort of demonstrates basic colored reflection and area lighting:
https://zalo.github.io/three.js/examples/#webgl_shadowmap_progressive (WIP)

This one is a lot more complicated than the Shadow PR, so it'll probably take me a bit longer to get it to look nice...

**Comment by mrdoob** [comment]
@zalo that runs pretty well on my cheap Android phone 😮

**Comment by EliasHasle** [comment]
My cheap Android phone shows it like this (with FPS eyeballed to ~10):
![_20210313_141212](https://user-images.githubusercontent.com/17208315/111031144-5d01c780-8406-11eb-85e7-25299b6606d8.JPG)

Do you have a reference screenshot for how it should look?

**Comment by zalo** [comment]
![0B594B05-A2F5-4FBF-AF58-3F17439EAF1E](https://user-images.githubusercontent.com/174475/111044000-ca0f6000-83fa-11eb-955b-beb99bf6b338.png)

Took this on a Pixel 4 (the banding isn’t visible on Desktops with full float precision buffers... though I’d like to fix that with additional dithering...).   @EliasHasle is there any chance you could check the console for errors and report them to me?  As far as I know, the most exotic things I’m doing are float buffers and writing depth in the fragment shader...

If the latter is the issue... it may be possible to jitter depth in vertex space instead...

**Comment by mrdoob** [comment]
@zalo That looks good to me. Should that become another example?

**Comment by bhouston** [comment]
Looks broken on Safari on M1 in WebGL2 mode.
![Screen Shot 2021-03-15 at 2 28 54 PM](https://user-images.githubusercontent.com/588541/111202965-d92d1400-859a-11eb-971f-61e50097492c.png)

**Comment by mrdoob** [comment]
@bhouston Could you try if it's also broken on Chrome on M1?

**Comment by bhouston** [comment]
It works fine in Chrome on M1: 

![Screen Shot 2021-03-15 at 2 39 47 PM](https://user-images.githubusercontent.com/588541/111204244-54db9080-859c-11eb-847a-d60ac5704d8d.png)

**Comment by mrdoob** [comment]
@bhouston Thanks for testing. Would you like to report the issue to Apple?

**Comment by gkjohnson** [comment]
This looks great -- here's an artifact I see in Chrome on Mac with a Radeon Pro 560 that I don't see in @bhouston's screenshot (notice the hard edge in the lighting):

![image](https://user-images.githubusercontent.com/734200/111205102-1934cc80-8584-11eb-9326-736c044a8557.png)

**Comment by carstenschwede** [comment]
Results for Macbook Pro 2020, Radeon 5600M. Hard edge in Chrome and Firefox, no edge in Safari.
<img width="1378" alt="Screenshot 2021-03-16 at 16 40 30" src="https://user-images.githubusercontent.com/234960/111337712-980a3200-8676-11eb-9f50-24f55397d0b7.png">

**Comment by zalo** [comment]
That artifact is very interesting; I think Radeon is the common element there.  My wildest guess is that negative orthographic near clipping planes aren't supported on Radeon graphics cards on Mac...?

Safari was acting broken on my Mid-2015 Intel Iris Macbook, but Chrome works flawlessly.  It turns out that is because `renderer.extensions.has( 'EXT_frag_depth' )` returns `true` on Mac Safari, but it's not true!  It's a lie!  It also runs much more slowly than Chrome, so I'm additionally downgrading Safari to mobile settings.

Do any of those changes help on Radeon Safari, @gkjohnson and @carstenschwede ?   If not, could you also hit the "Debug Lightmap" setting to see what that mysterious line looks like there?   Perhaps also try moving the lighting/object around to see if the line moves? Updated Here: https://zalo.github.io/three.js/examples/#webgl_shadowmap_progressive

With regards to a new example, I'd really like to get the code cleaned up, add more configurability, add more bounces, and increase the complexity of the example scene/("unit test") to make this a suitable addition to three.js.  [The radiosity component currently triples the surface area of the lightmapping script to ~700 lines](https://github.com/zalo/three.js/compare/e16c5e1...zalo:progressive-radiosity), so I'd like to par that down if possible...

**Comment by carstenschwede** [comment]
Looking more closely now, there are actually different edges in Safari and Chrome (Radeon 5600M, macOS). Edge does not move with lights/objects and is not present in early iterations.

Chrome:
<img width="978" alt="Screenshot 2021-03-17 at 01 00 31" src="https://user-images.githubusercontent.com/234960/111395057-42587880-86bc-11eb-97ed-9966785bbb1d.png">

Safari:
<img width="974" alt="Screenshot 2021-03-17 at 01 00 40" src="https://user-images.githubusercontent.com/234960/111395065-44223c00-86bc-11eb-9d4c-4b1a49beb928.png">

**Comment by carstenschwede** [comment]
Overriding software rendering list and enabling draft extensions in Chrome/macOS/Radeon also has no effect on the edge.

**Comment by carstenschwede** [comment]
Here is a montage of how the edge appears in later iterations (Chrome/MacOS/Radeon)
![montage](https://user-images.githubusercontent.com/234960/111494388-b471a180-873e-11eb-8748-6c424aacf23e.jpg)

**Comment by bhouston** [comment]
If this is in an example folder, rather than being a core feature, I wouldn't require it to be absolutely perfect on all devices in the first PR, because that could stall it from getting in.

**Comment by carstenschwede** [comment]
The light target is currently on top of the object. Would it make sense to add a corresponding helper to indicate the orientation of DirectionalLight? I assumed the control was for a PointLight and was irritated when the shadows were pointing towards it.
<img width="1532" alt="Screenshot 2021-03-17 at 17 46 17" src="https://user-images.githubusercontent.com/234960/111507588-1afcbc80-874b-11eb-8ea9-9cf4cab26282.png">

**Comment by EliasHasle** [comment]
> @EliasHasle is there any chance you could check the console for errors and report them to me? As far as I know, the most exotic things I’m doing are float buffers and writing depth in the fragment shader...

<details>
<summary>Beginning of log</summary>
MESSAGE: Safari: false

LEVEL: LOG TIMESTAMP: 14:28:37

LINE: 40 IN: https://zalo.github.io/three.js/examples/jsm/misc/ProgressiveLightMap.js 

MESSAGE: WebGL2: true

LEVEL: LOG TIMESTAMP: 14:28:37

LINE: 41 IN: https://zalo.github.io/three.js/examples/jsm/misc/ProgressiveLightMap.js 

MESSAGE: FragDepth: false

LEVEL: LOG TIMESTAMP: 14:28:37

LINE: 42 IN: https://zalo.github.io/three.js/examples/jsm/misc/ProgressiveLightMap.js 

MESSAGE: THREE.WebGLRenderer: EXT_color_buffer_float extension not supported.

LEVEL: WARNING TIMESTAMP: 14:28:41

LINE: 14820 IN: https://zalo.github.io/three.js/build/three.module.js 

MESSAGE: THREE.WebGLRenderer: EXT_color_buffer_float extension not supported.

LEVEL: WARNING TIMESTAMP: 14:28:44

LINE: 14820 IN: https://zalo.github.io/three.js/build/three.module.js 

MESSAGE: THREE.WebGLRenderer: EXT_color_buffer_float extension not supported.

LEVEL: WARNING TIMESTAMP: 14:28:46

LINE: 14820 IN: https://zalo.github.io/three.js/build/three.module.js 

MESSAGE: THREE.WebGLRenderer: EXT_color_buffer_float extension not supported.

LEVEL: WARNING TIMESTAMP: 14:28:47

LINE: 14820 IN: https://zalo.github.io/three.js/build/three.module.js 

MESSAGE: [.WebGL-0x7e3ee3b400]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glClear: framebuffer incomplete

LEVEL: WARNING TIMESTAMP: 14:28:47

LINE: 0 IN: https://zalo.github.io/three.js/examples/webgl_shadowmap_progressive.html 

MESSAGE: [.WebGL-0x7e3ee3b400]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glDrawElements: framebuffer incomplete

LEVEL: WARNING TIMESTAMP: 14:28:47

LINE: 0 IN: https://zalo.github.io/three.js/examples/webgl_shadowmap_progressive.html 

MESSAGE: [.WebGL-0x7e3ee3b400]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : GetShaderiv: 

LEVEL: WARNING TIMESTAMP: 14:28:47

LINE: 0 IN: https://zalo.github.io/three.js/examples/webgl_shadowmap_progressive.html 

MESSAGE: [.WebGL-0x7e3ee3b400]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glDrawElements: framebuffer incomplete
</details>

**Comment by arturitu** [comment]
@zalo if I export a simple mesh from Blender with a UV map it doesn't work as expected. It is all black. Do you know if it is needed to do something special with the UVs before export it? (I tried with importing other gltf files included on the examples folder like 'Parrot.glb' or 'Horse.glb' and it happens too) (Surely I'm missing something)

**Comment by zalo** [comment]
@arturitu [This script](https://github.com/mrdoob/three.js/blob/master/examples/js/misc/ProgressiveLightMap.js) uses the uv2 channel (primarily used for lightmapping?) by default.

**Comment by arturitu** [comment]
> It may make sense to copy the UVs from uv to uv2 if uv2 doesn't exist, but the uv2 channel has extra padding requirements in my naive implementation...
> 
Yes, I saw that and I think it makes sense because three.js I think only support 2 sets of UVs, but If I export from Blender a simple Icosphere (or any other mesh) with this UVs ...

![Screenshot 2022-01-20 at 05 22 12](https://user-images.githubusercontent.com/5744112/150272943-2bd97372-e70f-486e-88d2-5167f2a28b78.png)

... it looks like this:

![image](https://user-images.githubusercontent.com/5744112/150273166-f61c8544-9456-4b9f-8813-2db5c068fd4c.png)

If I export the same mesh with this other UVs ...

![image](https://user-images.githubusercontent.com/5744112/150273553-75a0f9ad-86c1-42a5-a73d-447acbed9435.png)

... it looks better but (I think) not all faces are properly calculated:

![image](https://user-images.githubusercontent.com/5744112/150273745-3c0ae2b9-518c-4c1c-9231-7a0e9ee64fbc.png)

Here the blend and the .glb files used with your code: [test-UVs.zip](https://github.com/mrdoob/three.js/files/7902341/test-UVs.zip)

BUT, if you add directly on you webgl_shadowmap_progressive a Mesh with a THREE.Icosahedron geometry it works as expected. I replicated it using the threejs editor, exporting the glb and seeing in blender what UVs it generated:

![image](https://user-images.githubusercontent.com/5744112/150276198-ec6f9df3-68cc-4693-a139-2ddbc576f100.png)

... and now, it looks as expected:

![image](https://user-images.githubusercontent.com/5744112/150276299-19e4ca67-18ec-430b-9f9c-6237348b916a.png)

And then I tried to unwrap like this on Blender using seams but it doesn't work as well. I suspect that it may be related to the order of the vertices of the geometry, that maybe your code expects them in a certain order and with other glbs the order in the UVs is more complex.

My goal is to find if is possible to use any kind of generated meshes with a properly UVs to be used with this global illumination method.

**Comment by mishagrin** [comment]
@arturitu
Any progress on this?

**Comment by arturitu** [comment]
> @arturitu Any progress on this?
Nope. I did not go any further along this line because I did not get a workflow that I could control for the project.

**Comment by ToJans** [comment]
> 

FYI this might also be a viable (intermediate? good enough?) route: https://github.com/0beqz/screen-space-reflections
![image](https://user-images.githubusercontent.com/155977/187623809-89ada30a-4759-477c-b21a-cb8fc8384c8f.png)

**Comment by zalo** [comment]
Wow; that looks great!

And any screen-space technique is going to be a lot less invasive than a lightmapping-space technique :-)

I'm a big fan of what people have been able to accomplish with velocity and reprojection buffers. 
It's even possible that the light+shadow jittering method could benefit from working in screenspace that way.

**Comment by PetterGs** [comment]
If anyone is curious how to use procedural geometry to make an atlas for progressive lightmaps. I used xatlas-three to unwrap and atlas-pack the uv2. The trick is to set `fixWinding` to `true` and `rotateCharts` to `false` in the xatlas options.

Here is a TypeScript example:
```
const logProgress = true;

export const loadXAtlasThree = async () => {
    if (_loaded)
        return;

    const onProgress = (mode: number, progress: number) => {
        logProgress && console.log(`UV unwrap state: ${ProgressCategory[mode]} ${progress}%`);
    }
    await unwrapper.loadLibrary(
        onProgress,
        'https://cdn.jsdelivr.net/npm/xatlasjs@0.1.0/dist/xatlas.wasm',
        'https://cdn.jsdelivr.net/npm/xatlasjs@0.1.0/dist/xatlas.js',
    );
    _loaded = true;
};

export const unwrapUv2 = async (geometries: BufferGeometry[]) => {
    await unwrapper.packAtlas(geometries, 'uv2' as any);
    return geometries;
}

unwrapper.chartOptions = {
    fixWinding: true,
    maxBoundaryLength: 0,
    maxChartArea: 0,
    maxCost: 2,
    maxIterations: 1,
    normalDeviationWeight: 2,
    normalSeamWeight: 4,
    roundnessWeight: 0.009999999776482582,
    straightnessWeight: 6,
    textureSeamWeight: 0.5,
    useInputMeshUvs: false,
}

unwrapper.packOptions = {
    bilinear: true,
    blockAlign: false,
    bruteForce: false,
    createImage: false,
    maxChartSize: 0,
    padding: 0,
    resolution: 0,
    rotateCharts: false,
    rotateChartsToAxis: true,
    texelsPerUnit: 0
}
```

**Comment by Mugen87** [comment]
Many of the suggested features have been implemented over the years. The new post processing stack of `WebGPURenderer` supports the important effects that were discussed in this issue namely GTAO, SSGI, SSR and TRAA. Many of the new FX effects also support temporal accumulation/filtering when using the TRAA.

The `ProgressiveLightMap` component works in both `WebGPURenderer` and `WebGLRenderer`.

The "old" post processing of `WebGLRenderer` only has partial support for these techniques like SSR or GTAO. Because of the dated architecture and missing TSL support, we are implementing new effects only for `WebGPURenderer`. Additional effects for `WebGLRenderer` can be used over third-party projects.

Please create new issues for concrete feature requests or if you encounter problems with one of the existing components.

**Comment by richtrr** [comment]
Hey, someone called me back from the 2000s and asked if the now had implemented global illumination in 3D Web.

I replied that I'd find out...

Has the radiocity algorithm for baking light into vertices already been implemented?

Is Lighttracer implemented?

Is automatic generation of lightmaps for statistical scenes implemented?

Is UV editing coordinates implemented in all possible situations?

Is everything that was in 3ds Max 20 years ago already implemented?

My friend from the past is asking these questions. And I join in.

**Comment by richtrr** [comment]
![Image](https://github.com/user-attachments/assets/7f15bb20-12a9-4d6c-9057-ac457be3ba34)

One second for global illumination calculation, 25 years for software

**Comment by richtrr** [comment]
Why is everyone so fixated on textures?!) You can make meshes!

**Comment by richtrr** [comment]
> Many of the suggested features have been implemented over the years. The new post processing stack of `WebGPURenderer` supports the important effects that were discussed in this issue namely GTAO, SSGI, SSR and TRAA. Many of the new FX effects also support temporal accumulation/filtering when using the TRAA.
> 
> The `ProgressiveLightMap` component works in both `WebGPURenderer` and `WebGLRenderer`.
> 
> The "old" post processing of `WebGLRenderer` only has partial support for these techniques like SSR or GTAO. Because of the dated architecture and missing TSL support, we are implementing new effects only for `WebGPURenderer`. Additional effects for `WebGLRenderer` can be used over third-party projects.
> 
> Please create new issues for concrete feature requests or if you encounter problems with one of the existing components.

Небо такое же голубое, как если бы ты не продался, прояви мужественость и сражайся!

**Comment by richtrr** [comment]
Не хочу критиковать, но именно на этом моменте я не смог промолчать. То есть развития рендеринга глобального освещения нет уже много лет и в любой ситуации, когда нужно отрендерить освещение клент привязан к вендору. Также когда нужно отредактировать 3d-модель по минимуму, тпа там материал поправить, или... -- это не нормально, нужно другое решенеие для общих задач, и уж точно для задач редактирования uv, материалов, структуры сцены, и ... расчета глобального освещения.

Потому что, как говорит мой друг из 2000, или ты можешь или нет.
